### DISCUSSION TITLE
codeql vscode plugin show result in `#select` page but not `alert` page

### DISCUSSION THREAD

**Discussion by KiruaLawliet** [github/securitylab#516] [Q&A] [answered]
evnironment:
CodeQL command-line toolchain release 2.7.3.
vscode Version: 1.63.2
codeql plugin ver：v1.5.9
OS: Windows_NT x64 10.0.19042

database of project A：compiled from openjdk8u which is a huge java project（about 18,800+ java files）
database of project B：compiled from project A with additional two java projects（rome1.0 and rhino:js:1.7R2）

After successfully compiling , I checked database directory of project B and seen all source files of three projects included in src.zip file.
However，when I run query script individually on each database, query result of porject B is strange.

<h3> result of project A </h3>

this result is exactually what I want.

![image](https://user-images.githubusercontent.com/16208969/147809426-714edf66-2e84-4a75-b10d-eb9a47a33a94.png)

<h3> result of project B </h3>

All path nodes in the result showed above are from porject A.  Thus,  it should come out to share the same part of  query result, 
but as you can see  below，no result in  `alert` page, 

![image](https://user-images.githubusercontent.com/16208969/147809475-8497e6ef-d8e7-44c7-be4f-42d0999e766b.png)

just one result in `#select` page and not in path-problem view

![image](https://user-images.githubusercontent.com/16208969/147809489-114ab3f0-9408-43e9-a7ec-930f38940655.png)

don't know why, here are my simplify query script which got `isAdditionalTaintStep` removed：
```sql

/**@kind path-problem */

import java
import semmle.code.java.dataflow.TaintTracking

import DataFlow::PathGraph

class DeserialFlowConfig extends TaintTracking::Configuration
{
    DeserialFlowConfig()
    {
        this = "DeserialFlowConfig" 
    }

    override predicate isSource(DataFlow::Node node) {
        node.asParameter().getCallable().hasQualifiedName("java.util", "HashMap", "hash")
        and node.asParameter().getPosition() = 0
    }

    override predicate isSink(DataFlow::Node node) {
        exists(MethodAccess ma|
            ma.getMethod().hasName("hashCode")
            and ma.getEnclosingCallable().hasName("hashCode")
            and ma.getAnArgument() = node.asExpr()
            and ma.getMethod().getDeclaringType().hasQualifiedName("java.net", "URLStreamHandler")
        )

    }
}


from DeserialFlowConfig config, DataFlow::PathNode source, DataFlow::PathNode sink
where config.hasFlowPath(source, sink)
select sink.getNode(), source, sink, "RESULT"


```

**Comment by smowton** [comment, marked as answer]
Possibly related: are the source or sink source locations outside the source root for project B? If so I know they can get dropped from path explanation generation.

**Comment by KiruaLawliet** [comment]
@smowton No, all path nodes of query result of project A are from jdk8u and also I'm sure that project B has all those source files included in src.zip file which locates in database directory.

![image](https://user-images.githubusercontent.com/16208969/148149585-ba92693c-28d9-4715-9f91-99cd96399d26.png)

In additionally, Run `isSource` predicate and `isSink` predicate individually by `Quick Evaluation` on Porject B can get the correct result：

![image](https://user-images.githubusercontent.com/16208969/148150518-d7e32130-e92c-4c2b-b210-0174870258f4.png)

**Reply by smowton** [reply]
Are you able to share your database B and a query that produces the desired result (does the simplified query you pasted above produce it?)

**Reply by KiruaLawliet** [reply]
@smowton Yes, the simplified query also produces that strange result but not normal path-problem view on `alert` page. 
I shared database B by Git Large File Storage (LFS). Plz check if you can access this repository and fetch the file `openjdk8u-mixed-database.zip` https://github.com/KiruaLawliet/TransferStation/  
also plz use this query：
```sql
/**@kind path-problem */

import java
import semmle.code.java.dataflow.TaintTracking

import DataFlow::PathGraph

class DeserialFlowConfig extends TaintTracking::Configuration
{
    DeserialFlowConfig()
    {
        this = "DeserialFlowConfig" 
    }

    override predicate isSource(DataFlow::Node node) {
        node.asParameter().getCallable().hasQualifiedName("java.util", "HashMap", "hash")
        and node.asParameter().getPosition() = 0
    }

    override predicate isSink(DataFlow::Node node) {
        exists(MethodAccess ma|
            ma.getMethod().hasName("hashCode")
            and ma.getEnclosingCallable().hasName("hash")
            and ma.getQualifier() = node.asExpr()
            and ma.getEnclosingCallable().getDeclaringType().hasQualifiedName("java.util", "HashMap")
        )

    }
}


from DeserialFlowConfig config, DataFlow::PathNode source, DataFlow::PathNode sink
where config.hasFlowPath(source, sink)
select sink.getNode(), source, sink, "RESULT"
```

and it's query result on my pc.

![image](https://user-images.githubusercontent.com/16208969/148324244-e85cfbd6-c6c2-4b1b-a88f-9d398983be0b.png)

build commands of database B：
```shell
codeql database create java-database --language=java --overwrite --command='build.sh'

# build.sh
cd ../openjdk8u/
make all
cd ../rhino-1.7.10/
mvn clean compile
cd ../rome1.0/
mvn clean compile
```

**Reply by smowton** [reply]
The cause is indeed down to source locations: it looks like you've run the build command from `/home/ming/software/build_jdk/build_common`, but the `select`ed element is `key` which is in `/home/ming/software/build_jdk/openjdk-jdk8u-master`. The result is dropped from path explanations because it isn't within `build_common`.

This could be overcome by phrasing your build process such that it commences in a path that includes all the code you wish to consider in scope. However we can also just quickly hack it: edit `java-database/codeql-database.yml` and `java-database/.dbinfo`, replacing `/home/ming/software/build_jdk/build_common` with `/home/ming/software/build_jdk`. Then all your code will be considered "in scope" and you'll get a path explanation as expected.

**Reply by KiruaLawliet** [reply]
@smowton  Many thanks !! It really works!!
